R/paste_sep_linter.R

```diff
@@ -0,0 +1,39 @@
+#' Block usage of paste() with sep=""
+#'
+#' [paste0()] is a faster, more concise alternative to using `paste(sep = "")`.
+#'
+#' @evalRd rd_tags("paste_sep_linter")
+#' @seealso [linters] for a complete list of linters available in lintr.
+#' @export
+paste_sep_linter <- function() {
+  Linter(function(source_file) {
+    if (length(source_file$xml_parsed_content) == 0L) {
+      return(list())
+    }
+
+    xml <- source_file$xml_parsed_content
+
+    paste_cond <- "expr[SYMBOL_FUNCTION_CALL[text() = 'paste']]"
+    sep_cond <- file.path(
+      "SYMBOL_SUB[text() = 'sep']",
+      # NB: length-2 means '' or "" (*not* 'ab' or 'cd' which have length 4)
+      "following-sibling::expr[1][STR_CONST[string-length(text()) = 2]]"
```

Comment:
Edge case: What about `r"()"`?
Probably easier to understand with explicit comparisons.

Reply:
yea... this is gonna come up a few times. we're still on R 3.6.3 so haven't had to lint for any raw strings stuff yet.

in the case of this linter, I think it's fine as is... `paste(sep = R'()')` seems unlikely.

this linter had mainly been useful for tidying up very old code written before paste0 existed, FWIW

PS we may have to do a thing where we merge first and tidy up later, for any case where there's something unique to do for R>4.0. it'll be tough to do such changes in my work environment.

Reply:
I can chip in with 4.x stuff.
It should also be safe and future proof to instead check `identical(eval(xml_text(sep_str_const)), "")`. WDYT?